Use an explicit unsigned comparison for span index checks

stephentyrone opened the pull request:
LLVM ought to be able to do this transformation for us, but it currently fails to do so. We can code around it easily enough.

stephentyrone committed:
Use an explicit unsigned comparison for span index checks

LLVM ought to be able to do this transformation for us, but it currently fails to do so. We can code around it easily enough.

meg-gupta reviewed (approved):
This seems to be the same underlying issue as rdar://154950810 for which @eeckstein has a PR - https://github.com/swiftlang/swift/pull/82804 

This change looks good as well especially if you plan to cherry-pick

glessard reviewed (approved):
<s>We should have the same changes in `RawSpan` and `MutableRawSpan`.</s>
I'd already written the bounds checks that way in `RawSpan` and `MutableRawSpan`, so no change needed there.

stephentyrone commented:
@meg-gupta right, hopefully we can land https://github.com/swiftlang/swift/pull/82804 and fix the problem more generally

stephentyrone committed:
Add file-check tests for bounds-check optimizations on Span

And apply the unsigned-compare optimization one other place that I missed the first time, as revealed by the new tests.

stephentyrone committed:
Adjust file check to reflect file renaming

stephentyrone commented:
@swift-ci smoke test

stephentyrone committed:
Disable Span bounds check elimination tests for asserts builds

The asserts introduce null pointer checks that make the test too fragile.

stephentyrone commented:
@swift-ci smoke test

stephentyrone commented:
@swift-ci smoke test

stephentyrone commented:
@swift-ci smoke test macOS
